### ISSUE TITLE
KafkaJS has UnhandledPromiseRejection somewhere in the code

### ISSUE THREAD

**Issue by goriunov** [tulios/kafkajs#702]
Was running stress tests on the kafkajs and picked up this issue (error does not matter most important part is UnhandledPromiseRejectionWarning):

```
(node:78646) UnhandledPromiseRejectionWarning: KafkaJSProtocolError: The coordinator is not aware of this member
    at createErrorFromCode (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/protocol/error.js:537:10)
    at Object.parse (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/protocol/requests/heartbeat/v0/response.js:20:11)
    at Connection.send (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/network/connection.js:317:35)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:78646) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2017)
```

Looks like `heartbeat` has unhandled promise or throwing inside async function somewhere

Was running latest `BETA` version  ( "kafkajs": "1.13.0-beta.15")

**Comment by goriunov** [comment]
You should be able to easily reproduce that by changing `heartbeat/v0/response.json` to always throw and run consumer.

```js
const parse = async data => {
  // if (failure(data.errorCode)) {
  throw createErrorFromCode()
  // }
  return data
}
```

Looks like it may be related to reties:

```
Passed here
[34mdebug[39m: Suppressed error:  {"timestamp":"2020-04-19T21:17:58.870Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":5,"size":10}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.870Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":5,"payload":{"type":"Buffer","data":"[filtered]"}}
Passed here
[34mdebug[39m: Suppressed error:  {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":6,"size":10}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":6,"payload":{"type":"Buffer","data":"[filtered]"}}
Passed here
[34mdebug[39m: Suppressed error:  {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":7,"size":10}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":7,"payload":{"type":"Buffer","data":"[filtered]"}}
Passed here
[34mdebug[39m: Suppressed error:  {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":8,"size":10}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":8,"payload":{"type":"Buffer","data":"[filtered]"}}
Passed here
[34mdebug[39m: Suppressed error:  {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":9,"size":10}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":9,"payload":{"type":"Buffer","data":"[filtered]"}}
Passed here
[34mdebug[39m: Suppressed error:  {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":10,"size":10}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":10,"payload":{"type":"Buffer","data":"[filtered]"}}
Passed here
[34mdebug[39m: Suppressed error:  {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":11,"size":10}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.871Z","logger":"kafkajs","message":"Response Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","error":"Unknown error code undefined","correlationId":11,"payload":{"type":"Buffer","data":"[filtered]"}}
Passed here
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.873Z","logger":"kafkajs","message":"Request Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","correlationId":16,"expectResponse":true,"size":121}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.873Z","logger":"kafkajs","message":"Request Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","correlationId":17,"expectResponse":true,"size":121}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.873Z","logger":"kafkajs","message":"Request Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","correlationId":18,"expectResponse":true,"size":121}
[34mdebug[39m: {"timestamp":"2020-04-19T21:17:58.874Z","logger":"kafkajs","message":"Request Heartbeat(key: 12, version: 1)","broker":"xyz:9092","clientId":"kafkajs-87503","correlationId":19,"expectResponse":true,"size":121}
(node:87503) UnhandledPromiseRejectionWarning: KafkaJSProtocolError: Unknown error code undefined
    at createErrorFromCode (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/protocol/error.js:537:10)
    at Object.parse (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/protocol/requests/heartbeat/v0/response.js:20:9)
    at Connection.send (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/network/connection.js:317:35)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:87503) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 17)
(node:87503) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by Nevon** [comment]
I cannot seem to reproduce this. See https://github.com/tulios/kafkajs/tree/repro-702.

I get into the regular crash handling:

```
$ node examples/consumer.js
{"level":"ERROR","timestamp":"2020-04-20T06:33:09.366Z","logger":"kafkajs","message":"[Connection] Response Metadata(key: 3, version: 5)","broker":"192.168.0.17:9100","clientId":"example-consumer","error":"There is no leader for this topic-partition as we are in the middle of a leadership election","correlationId":3,"size":126}
Attempt to reproduce #702
{"level":"INFO","timestamp":"2020-04-20T06:33:09.692Z","logger":"kafkajs","message":"[Consumer] Starting","groupId":"test-group"}
{"level":"ERROR","timestamp":"2020-04-20T06:33:09.789Z","logger":"kafkajs","message":"[Connection] Response GroupCoordinator(key: 10, version: 1)","broker":"localhost:9100","clientId":"example-consumer","error":"The group coordinator is not available","correlationId":2,"size":55}
{"level":"ERROR","timestamp":"2020-04-20T06:33:10.140Z","logger":"kafkajs","message":"[Connection] Response GroupCoordinator(key: 10, version: 1)","broker":"localhost:9100","clientId":"example-consumer","error":"The group coordinator is not available","correlationId":3,"size":55}
{"level":"ERROR","timestamp":"2020-04-20T06:33:10.967Z","logger":"kafkajs","message":"[Connection] Response GroupCoordinator(key: 10, version: 1)","broker":"localhost:9100","clientId":"example-consumer","error":"The group coordinator is not available","correlationId":4,"size":55}
{"level":"INFO","timestamp":"2020-04-20T06:33:12.761Z","logger":"kafkajs","message":"[Runner] Consumer has joined the group","groupId":"test-group","memberId":"example-consumer-2a9eaf3e-97cc-4cbc-839c-47550bc88f97","leaderId":"example-consumer-2a9eaf3e-97cc-4cbc-839c-47550bc88f97","isLeader":true,"memberAssignment":{"topic-test":[0]},"groupProtocol":"RoundRobinAssigner","duration":3067}
{"level":"ERROR","timestamp":"2020-04-20T06:33:17.851Z","logger":"kafkajs","message":"[Connection] Response Heartbeat(key: 12, version: 1)","broker":"localhost:9094","clientId":"example-consumer","error":"Unknown error code 0","correlationId":5,"size":10}
{"level":"ERROR","timestamp":"2020-04-20T06:33:17.852Z","logger":"kafkajs","message":"[Consumer] Crash: KafkaJSNumberOfRetriesExceeded: Unknown error code 0","groupId":"test-group","retryCount":0,"stack":"KafkaJSNonRetriableError\n  Caused by: KafkaJSProtocolError: Unknown error code 0\n    at createErrorFromCode (/Users/tommybrunn/workspace/kafkajs/src/protocol/error.js:537:10)\n    at Object.parse (/Users/tommybrunn/workspace/kafkajs/src/protocol/requests/heartbeat/v0/response.js:20:9)\n    at Connection.send (/Users/tommybrunn/workspace/kafkajs/src/network/connection.js:317:35)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)\n    at async Broker.heartbeat (/Users/tommybrunn/workspace/kafkajs/src/broker/index.js:314:12)\n    at async ConsumerGroup.heartbeat (/Users/tommybrunn/workspace/kafkajs/src/consumer/consumerGroup.js:317:7)\n    at async Runner.fetch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:369:5)\n    at async /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:384:9"}
{"level":"INFO","timestamp":"2020-04-20T06:33:17.908Z","logger":"kafkajs","message":"[Consumer] Stopped","groupId":"test-group"}
{"level":"ERROR","timestamp":"2020-04-20T06:33:17.909Z","logger":"kafkajs","message":"[Consumer] Restarting the consumer in 297ms","retryCount":0,"retryTime":297,"groupId":"test-group"}
```

**Comment by ankon** [comment]
We have seen the same once as well, didn't get around yet to filing this :(

~~~~
(node:17) UnhandledPromiseRejectionWarning: KafkaJSProtocolError: The group is rebalancing, so a rejoin is needed
     at createErrorFromCode (/app/node_modules/kafkajs/src/protocol/error.js:537:10)
     at Object.parse (/app/node_modules/kafkajs/src/protocol/requests/heartbeat/v0/response.js:20:11)
     at Connection.send (/app/node_modules/kafkajs/src/network/connection.js:317:35)
     at runMicrotasks (<anonymous>)
     at processTicksAndRejections (internal/process/task_queues.js:97:5)
     at Broker.heartbeat (/app/node_modules/kafkajs/src/broker/index.js:314:12)
     at ConsumerGroup.heartbeat (/app/node_modules/kafkajs/src/consumer/consumerGroup.js:317:7)
     at /app/node_modules/kafkajs/src/consumer/runner.js:346:15
 (node:17) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 529519)
 (node:17) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
~~~~

**Comment by goriunov** [comment]
@Nevon just tried running from your commit. Looks like there is one more condition which is necessary to reproduce the bug. It should constantly consume messages from the topic.

So i have tweaked producer and started publishing messages to the topic:
```
const { Kafka, logLevel } = require('../index')

const kafka = new Kafka({
  logLevel: logLevel.DEBUG,
  brokers: [`xyz:9092`],
  clientId: 'example-producer',
})

const topic = 'topic-test'
const producer = kafka.producer()

const getRandomNumber = () => Math.round(Math.random(10) * 1000)
const createMessage = num => ({
  key: `key-${num}`,
  value: `value-${num}-${new Date().toISOString()}`,
})

const sendMessage = () => {
  return producer
    .send({
      topic,
      messages: Array(getRandomNumber())
        .fill()
        .map(_ => createMessage(getRandomNumber())),
    })
    .then(console.log)
    .catch(e => console.error(`[example/producer] ${e.message}`, e))
}

const run = async () => {
  await producer.connect()
  setInterval(sendMessage, 100)
}

run().catch(e => console.error(`[example/producer] ${e.message}`, e))

const errorTypes = ['unhandledRejection', 'uncaughtException']
const signalTraps = ['SIGTERM', 'SIGINT', 'SIGUSR2']

errorTypes.map(type => {
  process.on(type, async () => {
    try {
      console.log(`process.on ${type}`)
      await producer.disconnect()
      process.exit(0)
    } catch (_) {
      process.exit(1)
    }
  })
})

signalTraps.map(type => {
  process.once(type, async () => {
    try {
      await producer.disconnect()
    } finally {
      process.kill(process.pid, type)
    }
  })
})

```

Then run consumer and in some time it crashed with unhandled rejection...

**Comment by Nevon** [comment]
Again, I'm not getting an unhandled promise rejection. The error is caught and handled by the crash handler:

```
- topic-test[0 | 72119] / 1587378593021 key-84#value-84-2020-04-20T10:29:53.018Z
{"level":"ERROR","timestamp":"2020-04-20T10:29:53.037Z","logger":"kafkajs","message":"[Connection] Response Heartbeat(key: 12, version: 1)","broker":"localhost:9097","clientId":"example-consumer","error":"Unknown error code 0","correlationId":62,"size":10}
{"level":"ERROR","timestamp":"2020-04-20T10:29:53.037Z","logger":"kafkajs","message":"[Consumer] Crash: KafkaJSNumberOfRetriesExceeded: Unknown error code 0","groupId":"test-group","retryCount":0,"stack":"KafkaJSNonRetriableError\n  Caused by: KafkaJSProtocolError: Unknown error code 0\n    at createErrorFromCode (/Users/tommybrunn/workspace/kafkajs/src/protocol/error.js:537:10)\n    at Object.parse (/Users/tommybrunn/workspace/kafkajs/src/protocol/requests/heartbeat/v0/response.js:20:9)\n    at Connection.send (/Users/tommybrunn/workspace/kafkajs/src/network/connection.js:317:35)\n    at processTicksAndRejections (internal/process/task_queues.js:97:5)\n    at async Broker.heartbeat (/Users/tommybrunn/workspace/kafkajs/src/broker/index.js:314:12)\n    at async ConsumerGroup.heartbeat (/Users/tommybrunn/workspace/kafkajs/src/consumer/consumerGroup.js:317:7)\n    at async Runner.processEachMessage (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:181:7)\n    at async onBatch (/Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:298:9)\n    at async /Users/tommybrunn/workspace/kafkajs/src/consumer/runner.js:345:15"}
{"level":"INFO","timestamp":"2020-04-20T10:29:53.049Z","logger":"kafkajs","message":"[Consumer] Stopped","groupId":"test-group"}
{"level":"ERROR","timestamp":"2020-04-20T10:29:53.049Z","logger":"kafkajs","message":"[Consumer] Restarting the consumer in 303ms","retryCount":0,"retryTime":303,"groupId":"test-group"}
```

**Comment by goriunov** [comment]
@Nevon Looks like unhandled rejection is not triggered on every run i think it is due to `heartbeatInterval` so i have change `heartbeatInterval` to 0 it also is important to have messages in the topic (probably good to have them on different partitions) after running few times looks like error appears more often than not (sometimes need to wait abit on each run). Also i have gone in to runner and on each place where we call `await heartbeat()` did a fake `await new Promise((resolve, reject) => reject(new Error('Did not work')))` and looks like in some of the places it fails with unhandled rejection...

**Comment by goriunov** [comment]
During my chaos testing was able to pull another unhanded rejection 

```
(node:17090) UnhandledPromiseRejectionWarning: KafkaJSConnectionError: Not connected
    at Connection.failIfNotConnected (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/network/connection.js:352:13)
    at Connection.send (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/network/connection.js:275:10)
    at Broker.heartbeat (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/broker/index.js:314:34)
    at ConsumerGroup.heartbeat (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/consumer/consumerGroup.js:317:30)
    at concurrently (/Users/dmitrii/Desktop/Projects/at_connector_common/node_modules/kafkajs/src/consumer/runner.js:346:40)
    at process._tickCallback (internal/process/next_tick.js:68:7)
(node:17090) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 169)
```

**Comment by goriunov** [comment]
Just run version 1.12 and it works properly... Looks like it was introduced in some latest things...

**Comment by goriunov** [comment]
@Nevon after debuging abit more have found out that this like causes some issues
https://github.com/tulios/kafkajs/blob/2faef0719eeba50759eccc2b7d8800dbe63803f3/src/consumer/runner.js#L348

Looks like it is throwing in to this lock which has not yet been awaited which means it will throw unhandled

https://github.com/goriunov/kafkajs/blob/c76c34663fbb53f8df419e925c901d98d6a6ff4e/src/consumer/runner.js#L318

**Comment by goriunov** [comment]
@Nevon @tulios  found simpler way to reproduce issue. Change consumer to be something like that:

```js
const run = async () => {
  await consumer.connect()
  await consumer.subscribe({ topic, fromBeginning: true })
  await consumer.run({
    partitionsConsumedConcurrently: 2,
    eachMessage: async ({ topic, partition, message }) => {
      throw new Error('Did not work')
    },
  })
}
```

Before running consumer it is important to have messages in multiple partitions and probably good to have quite abit of messages so pre publish some 😄 

After running consumer it may work correctly for abit and than will fail with `Unhandled promise rejection` (due to parallel processing and not handling lock properly)

Please let me know if you can observe the issue now...

**Comment by WaleedAshraf** [comment]
With the last example, for me, it reaches `KafkaJSNumberOfRetriesExceeded` and then restarts.

**Comment by ankon** [comment]
The code in question has seen quite some changes in the current unreleased version, so that would somewhat be expected I'd say.

**Comment by goriunov** [comment]
@WaleedAshraf yeah i could not see that in 1.12 only on latest master can you try running on master and see if you can hit that bug.... With above example and few thousands messages in the topic and partitions i can reproduce every single time

**Comment by WaleedAshraf** [comment]
Tested "kafkajs": "1.13.0-beta.15" but was not able to reproduce the issue.
I tested with some already published messages, by just consuming the topic like above. Also, I have a lot of partitions for one topic, maybe that makes some difference.

**Comment by goriunov** [comment]
@WaleedAshraf interesting my one is always crashing may be i have different setup.....

**Comment by goriunov** [comment]
@WaleedAshraf this logs is just a fork of `master` (latest) with only change in consumer example:

```
-> kafkajs dmitrii$ node examples/consumer.js 
{"level":"INFO","timestamp":"2020-04-22T21:42:13.784Z","logger":"kafkajs","message":"[Consumer] Starting","groupId":"test-group"}
{"level":"INFO","timestamp":"2020-04-22T21:42:14.262Z","logger":"kafkajs","message":"[Runner] Consumer has joined the group","groupId":"test-group","memberId":"example-consumer-e8df5ed4-cd1b-47d4-9859-d1d3c9efc4e6","leaderId":"example-consumer-e8df5ed4-cd1b-47d4-9859-d1d3c9efc4e6","isLeader":true,"memberAssignment":{"xyz":[0,1,10,11,2,3,4,5,6,7,8,9]},"groupProtocol":"RoundRobinAssigner","duration":475}
{"level":"ERROR","timestamp":"2020-04-22T21:42:17.877Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"xyz","partition":1,"offset":"155857815","stack":"Error: Did not work\n    at Runner.eachMessage (/Users/dmitrii/Desktop/github/kafkajs/examples/consumer.js:20:13)\n    at Runner.processEachMessage (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:164:20)\n    at onBatch (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:298:20)\n    at concurrently (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:345:21)\n    at invoke (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:51:7)\n    at Promise (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:60:53)\n    at new Promise (<anonymous>)\n    at action (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:60:20)\n    at batches.map.batch (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:335:11)"}
{"level":"ERROR","timestamp":"2020-04-22T21:42:17.878Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"xyz","partition":10,"offset":"197257578","stack":"Error: Did not work\n    at Runner.eachMessage (/Users/dmitrii/Desktop/github/kafkajs/examples/consumer.js:20:13)\n    at Runner.processEachMessage (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:164:20)\n    at onBatch (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:298:20)\n    at concurrently (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:345:21)\n    at invoke (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:51:7)\n    at Promise (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:60:53)\n    at new Promise (<anonymous>)\n    at action (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:60:20)\n    at batches.map.batch (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:335:11)"}
{"level":"ERROR","timestamp":"2020-04-22T21:42:17.879Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"xyz","partition":4,"offset":"177473653","stack":"Error: Did not work\n    at Runner.eachMessage (/Users/dmitrii/Desktop/github/kafkajs/examples/consumer.js:20:13)\n    at Runner.processEachMessage (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:164:20)\n    at onBatch (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:298:20)\n    at concurrently (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:345:21)\n    at invoke (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:38:5)\n    at waiting.push.override (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:55:9)\n    at next (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:30:7)\n    at action.then.result (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:41:9)\n    at process._tickCallback (internal/process/next_tick.js:68:7)"}
{"level":"ERROR","timestamp":"2020-04-22T21:42:17.879Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"xyz","partition":7,"offset":"177816988","stack":"Error: Did not work\n    at Runner.eachMessage (/Users/dmitrii/Desktop/github/kafkajs/examples/consumer.js:20:13)\n    at Runner.processEachMessage (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:164:20)\n    at onBatch (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:298:20)\n    at concurrently (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:345:21)\n    at invoke (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:38:5)\n    at waiting.push.override (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:55:9)\n    at next (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:30:7)\n    at action.then.result (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:41:9)\n    at process._tickCallback (internal/process/next_tick.js:68:7)"}
process.on unhandledRejection
Unhandled promise rejection { error: 'Did not work',
  stack:
   'Error: Did not work\n    at Runner.eachMessage (/Users/dmitrii/Desktop/github/kafkajs/examples/consumer.js:20:13)\n    at Runner.processEachMessage (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:164:20)\n    at onBatch (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:298:20)\n    at concurrently (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:345:21)\n    at invoke (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:38:5)\n    at push (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:51:7)\n    at Promise (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:60:53)\n    at new Promise (<anonymous>)\n    at action (/Users/dmitrii/Desktop/github/kafkajs/src/utils/concurrency.js:60:20)\n    at batches.map.batch (/Users/dmitrii/Desktop/github/kafkajs/src/consumer/runner.js:335:11)' }
(node:99878) PromiseRejectionHandledWarning: Promise rejection was handled asynchronously (rejection id: 7)
{"level":"INFO","timestamp":"2020-04-22T21:42:19.940Z","logger":"kafkajs","message":"[Consumer] Stopped","groupId":"test-group"}

```

**Comment by WaleedAshraf** [comment]
Yes, I might be a bug. Maybe I missed something when trying to reproduce. I Will ping here if I catch this error.

**Comment by goriunov** [comment]
This is fixed #714
